### PR TITLE
fix(install): abort installation using git in active git session

### LINKED ISSUE #5353: :TSInstall while committing may break current git repo
### Describe the bug

Running `:TSInstall gitcommit` while committing in git broke my repo.
I use `require("nvim-treesitter.install").prefer_git = true`.

### To Reproduce

After encountering it on remote server, I immediately retried and reproduced it once on my home PC, but on subsequent attempts I can't reproduce it again.
Rough steps:
```
mkdir ~/Downloads/test
cd ~/Downloads/test
git init
touch asdf
git add asdf
git commit --no-gpg-sign
nvim asdf # ijklsdfjklsd:x
git add asdf
git commit # ikljsadflkjsfadjlk<ESC>:TSInstall gitcommit:x
git status
# > see error "fatal: unable to read 11544a188ec3192eb3aab635520020d5f175bd0b"
```

### Expected behavior

_No response_

### Output of `:checkhealth nvim-treesitter`

Initial encounter:
```
==============================================================================
nvim-treesitter: require("nvim-treesitter.health").check()

Installation ~
- WARNING `tree-sitter` executable not found (parser generator, only needed for :TSInstallFromGrammar, not required for :TSInstall)
- OK `node` found v12.22.9 (only needed for :TSInstallFromGrammar)
- OK `git` executable found.
- OK `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
  Version: cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
- OK Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

OS Info:
{
  machine = "x86_64",
  release = "6.2.0-31-generic",
  sysname = "Linux",
  version = "#31~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Wed Aug 16 13:45:26 UTC 2"
} ~

Parser/Features         H L F I J
  - bash                ✓ ✓ ✓ . ✓
  - c                   ✓ ✓ ✓ ✓ ✓
  - cmake               ✓ . ✓ ✓ .
  - cpp                 ✓ ✓ ✓ ✓ ✓
  - lua                 ✓ ✓ ✓ ✓ ✓
  - markdown            ✓ . ✓ ✓ ✓
  - markdown_inline     ✓ . . . ✓
  - python              ✓ ✓ ✓ ✓ ✓
  - query               ✓ ✓ ✓ ✓ ✓
  - sql                 ✓ . . ✓ ✓
  - vim                 ✓ ✓ ✓ . ✓
  - vimdoc              ✓ . . . ✓

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang} ~
```

Reproduced:
```
==============================================================================
nvim-treesitter: require("nvim-treesitter.health").check()

Installation ~
- WARNING `tree-sitter` executable not found (parser generator, only needed for :TSInstallFromGrammar, not required for :TSInstall)
- OK `node` found v20.6.0 (only needed for :TSInstallFromGrammar)
- OK `git` executable found.
- OK `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
  Version: cc (GCC) 13.2.1 20230801
- OK Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

OS Info:
{
  machine = "x86_64",
  release = "6.4.12-arch1-1",
  sysname = "Linux",
  version = "#1 SMP PREEMPT_DYNAMIC Thu, 24 Aug 2023 00:38:14 +0000"
} ~

Parser/Features         H L F I J
  - c                   ✓ ✓ ✓ ✓ ✓
  - cpp                 ✓ ✓ ✓ ✓ ✓
  - glsl                ✓ ✓ ✓ ✓ ✓
  - java                ✓ ✓ ✓ ✓ ✓
  - javascript          ✓ ✓ ✓ ✓ ✓
  - json5               ✓ . . . ✓
  - lua                 ✓ ✓ ✓ ✓ ✓
  - python              ✓ ✓ ✓ ✓ ✓
  - rust                ✓ ✓ ✓ ✓ ✓
  - toml                ✓ ✓ ✓ ✓ ✓
  - xml                 ✓ . ✓ ✓ ✓
  - yaml                ✓ ✓ ✓ ✓ ✓

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang} ~
```

### Output of `nvim --version`

Initial encounter:
```text
NVIM v0.10.0-dev-1031+gc431d820e
Build type: RelWithDebInfo
LuaJIT 2.1.1693350652
Run "nvim -V1 -v" for more info
```

Reproduced:
```
NVIM v0.9.1
Build type: Release
LuaJIT 2.1.1692616192

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

### Additional context

To fix the repo after that, I had to do a git hard reset (soft or mixed resets didn't work).

git errors looked like that:
```
> git commit
error: invalid object 100644 11544a188ec3192eb3aab635520020d5f175bd0b for '.github/workflows/ci.yml'
fatal: unable to read 11544a188ec3192eb3aab635520020d5f175bd0b
> git status
fatal: unable to read 11544a188ec3192eb3aab635520020d5f175bd0b
```
(named file is from [tree-sitter-gitcommit](https://github.com/gbprod/tree-sitter-gitcommit))
(for experiment, I created it in the root of my repo, after which it started complaining about the next file in tree-sitter-gitcommit).

The test git repo where I reproduced the bug: [test.zip](https://github.com/nvim-treesitter/nvim-treesitter/files/12547471/test.zip)

### PR DESCRIPTION
**Pull request by clason**
Problem: Installing a parser using git in an active git session (e.g., when editing a commit message) can corrupt the corresponding repository.

Solution: Check for typical environment variables first and abort installation if found.

Closes #5353

### CODE REVIEW

File: lua/nvim-treesitter/shell_command_selectors.lua [hunk 1]
```diff
@@ -289,6 +289,31 @@ function M.select_download_commands(repo, project_name, cache_folder, revision,
     local git_folder = utils.join_path(cache_folder, project_name)
     local clone_error = "Error during download, please verify your internet connection"
 
+    -- Running `git clone` or `git checkout` while running under Git (such as
+    -- editing a `git commit` message) will likely fail to install parsers
+    -- (such as 'gitcommit') and can also corrupt the index file of the current
+    -- Git repository. Check for typical git environemnt variables and abort if found.
```
Comment (rmacklin) [inline lua/nvim-treesitter/shell_command_selectors.lua:295]: I ran into this issue a couple weeks ago and made a note to look into it later, only to find this PR! Thank you for fixing this - getting a corrupt index file was quite surprising, so I'm glad that's resolved.

I noticed a minor typo in this comment which I've fixed in https://github.com/nvim-treesitter/nvim-treesitter/pull/6986

**Comment by rmacklin** [comment]
I'm also wondering if this PR means that https://github.com/nvim-treesitter/nvim-treesitter/pull/6761 can be reverted/dropped from the `main` branch.

**Comment by clason** [comment]
No. Dropping `git` support completely is the real solution; this is just a stopgap until 1.0 is released (from `main`).

**Comment by rmacklin** [comment]
So, is there no concern that dropping git support would break the setup for the folks that used `prefer_git = true` because of issues when using `curl` (https://github.com/nvim-treesitter/nvim-treesitter/issues/3232)?

**Comment by clason** [comment]
There are no issues using curl (if installed correctly).

**Comment by rmacklin** [comment]
It sounded like the issues were somewhat common, though, given the comments and reactions in that thread. But understood if that's not enough of a concern to reconsider dropping git support.

**Comment by clason** [comment]
The original reason to add `git` support were symlinks in parser repos. Before dropping support, I have gone through all supported parsers and made sure that they did not contain symlinks (or have them removed).

**Pull request merged**